### PR TITLE
Stamp detachedFieldIndex encoding with a version

### PR DESCRIPTION
**Pull request by daesunp**
## Description

This PR adds a version stamp to detachedFieldIndex blobs, in an effort to get all of the sharedTree's summarizables to be version stamped.

### CODE REVIEW

**Comment by Abe27342** [comment]
no need to include "The version stamp for ForestSummarizer will be included in this https://github.com/microsoft/FluidFramework/pull/17602." in the description. (but good to know, maybe an immediate comment on the PR would have been more appropriate)

File: experimental/dds/tree2/src/test/tree/detachedFieldIndex.spec.ts [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+						idAllocatorFromMaxId() as IdAllocator<ForestRootId>,
+						{
+							jsonValidator: typeboxValidator,
+						},
+					);
+					detachedFieldIndex.loadData(data);
+				});
+			}
+		});
+		describe("throws malformed data", () => {
```
Comment (Abe27342) [inline experimental/dds/tree2/src/test/tree/detachedFieldIndex.spec.ts:92]:
```suggestion
		describe("throws on receiving malformed data", () => {
```

File: experimental/dds/tree2/src/core/tree/detachedFieldIndex.ts [hunk 2]
```diff
@@ -141,29 +157,105 @@ export class DetachedFieldIndex {
...
 		this.rootIdAllocator = idAllocatorFromMaxId(
-			detachedFieldIndex.id,
+			detachedFieldIndex.maxId,
 		) as IdAllocator<ForestRootId>;
+		this.detachedNodeToField = detachedFieldIndex.data;
 	}
 }
+
+// Define the Major and Minor types:
+const MajorSchema = Type.Union([Type.String(), Type.Number(), Type.Undefined()]);
```
Comment (Abe27342) [inline experimental/dds/tree2/src/core/tree/detachedFieldIndex.ts:183]: Could you organize the format types and codec into their own file? Structure should match e.g. how modular change works:

`detachedFieldIndexFormat` for typebox schema declarations / encoded format types
`detachedFieldIndexTypes` for in-memory types (this is probably just the type `{
			data: NestedMap<Major, Minor, ForestRootId>;
			maxId: ForestRootId;
		}`, which you should name e.g. `DetachedFieldSummaryData`)
`detachedFieldIndexCodec` for the codec implementation

File: experimental/dds/tree2/src/core/tree/detachedFieldIndex.ts [hunk 3]
```diff
@@ -141,29 +157,105 @@ export class DetachedFieldIndex {
...
+const ForestRootIdSchema = brandedNumberType<ForestRootId>();
+
+// Define the tuple:
+const TupleSchema = Type.Tuple([MajorSchema, MinorSchema, ForestRootIdSchema]);
+
+export const Format = Type.Object({
+	version: Type.Literal(version),
+	data: Type.Array(TupleSchema),
+	maxId: ForestRootIdSchema,
+});
```
Comment (Abe27342) [inline experimental/dds/tree2/src/core/tree/detachedFieldIndex.ts:195]:
```suggestion
}, { additionalProperties: false });
```

File: experimental/dds/tree2/src/core/tree/detachedFieldIndex.ts [hunk 4]
```diff
@@ -141,29 +157,105 @@ export class DetachedFieldIndex {
...
+	version: Type.Number(),
+});
+type Versioned = Static<typeof Versioned>;
+
+export function makeDetachedNodeToFieldCodec({
+	jsonValidator: validator,
+}: ICodecOptions): IJsonCodec<{
+	data: NestedMap<Major, Minor, ForestRootId>;
+	maxId: ForestRootId;
+}> {
```
Comment (Abe27342) [inline experimental/dds/tree2/src/core/tree/detachedFieldIndex.ts:209]:
```suggestion
}, string> {
```

this should let you avoid casting above.

File: experimental/dds/tree2/src/test/tree/detachedFieldIndex.spec.ts (hunk 1 above)
Comment (daesunp) [reply experimental/dds/tree2/src/test/tree/detachedFieldIndex.spec.ts]: updated test description

File: experimental/dds/tree2/src/core/tree/detachedFieldIndex.ts (hunk 2 above)
Comment (daesunp) [reply experimental/dds/tree2/src/core/tree/detachedFieldIndex.ts]: Organized format types into different files

File: experimental/dds/tree2/src/core/tree/detachedFieldIndex.ts (hunk 3 above)
Comment (daesunp) [reply experimental/dds/tree2/src/core/tree/detachedFieldIndex.ts]: added suggested change

File: experimental/dds/tree2/src/core/tree/detachedFieldIndex.ts (hunk 4 above)
Comment (daesunp) [reply experimental/dds/tree2/src/core/tree/detachedFieldIndex.ts]: updated codec type and removed unnecessary casting

File: experimental/dds/tree2/src/core/tree/detachedFieldIndexTypes.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+ * Licensed under the MIT License.
+ */
+
+import { NestedMap } from "../../util";
+import { ForestRootId } from "./detachedFieldIndex";
+
+export type Major = string | number | undefined;
+export type Minor = number;
+
+export interface DetachedFieldSummaryType {
```
Comment (Abe27342) [inline experimental/dds/tree2/src/core/tree/detachedFieldIndexTypes.ts:12]:
```suggestion
export interface DetachedFieldSummaryData {
```
is a bit more consistent with EditManager's naming conventions. obviously suggestion will require renames elsewhere  so can't be applied as-is :)

**Review by Abe27342 — APPROVED** [review]

File: experimental/dds/tree2/src/core/tree/detachedFieldIndexTypes.ts (hunk 5 above)
Comment (daesunp) [reply experimental/dds/tree2/src/core/tree/detachedFieldIndexTypes.ts]: Updated interface name :)

**Review by Abe27342 — APPROVED** [review]

**Pull request merged**